contracts/SlashIndicator.sol

```diff
@@ -213,27 +215,30 @@ contract SlashIndicator is ISlashIndicator,System,IParamSubscriber, IApplication
       _evidence.voteA.tarNum == _evidence.voteB.tarNum, "no violation of vote rules");
 
     // BLS verification
+    require(verifyBLSSignature(_evidence.voteA, _evidence.voteAddr) &&
+      verifyBLSSignature(_evidence.voteB, _evidence.voteAddr), "verify signature failed");
+
+    // reward sender and felony validator if validator found
     (address[] memory vals, bytes[] memory voteAddrs) = IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).getLivingValidators();
-    address valAddr;
-    bytes memory voteAddr = _evidence.voteAddr;
     for (uint i; i < voteAddrs.length; ++i) {
-      if (BytesLib.equal(voteAddrs[i],  voteAddr)) {
-        valAddr = vals[i];
+      if (BytesLib.equal(voteAddrs[i],  _evidence.voteAddr)) {
+        uint256 amount = (address(SYSTEM_REWARD_ADDR).balance * finalitySlashRewardRatio) / 100;
+        ISystemReward(SYSTEM_REWARD_ADDR).claimRewards(msg.sender, amount);
+        IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).felony( vals[i]);
```

Comment:
In the common case, both `felony` and `VoteSlash` will trigger, so two cross chain package will be trigger, I think this need to optimized.